### PR TITLE
Support `std` aggregation in reduction

### PR DESCRIPTION
**Pull request by ttnghia**
This adds support for `std` (i.e., `stddev_samp`) aggregation in reduction context, completing https://github.com/NVIDIA/spark-rapids/issues/62.

Closes https://github.com/NVIDIA/spark-rapids/issues/7891 and closes https://github.com/NVIDIA/spark-rapids/issues/62.

### CODE REVIEW

**Comment by ttnghia** [comment]
Benchmark time running on a column of 1M rows randomly generated:

# CPU:
```
scala> val result = df.selectExpr("std(value)")
result: org.apache.spark.sql.DataFrame = [std(CAST(value AS DOUBLE)): double]

scala> spark.time(result.collect)
Time taken: 1780 ms                                                             
res0: Array[org.apache.spark.sql.Row] = Array([2888.906172169368])

scala> spark.time(result.collect)
Time taken: 40 ms
res1: Array[org.apache.spark.sql.Row] = Array([2888.906172169368])

scala> spark.time(result.collect)
Time taken: 38 ms
res2: Array[org.apache.spark.sql.Row] = Array([2888.906172169368])

....
```

# CPU with 1 partition:
```
scala> val result = df.repartition(1).selectExpr("std(value)")
result: org.apache.spark.sql.DataFrame = [std(CAST(value AS DOUBLE)): double]

scala> spark.time(result.collect)
Time taken: 682 ms
res3: Array[org.apache.spark.sql.Row] = Array([2888.906172169392])

scala> spark.time(result.collect)
Time taken: 146 ms
res4: Array[org.apache.spark.sql.Row] = Array([2888.906172169392])

scala> spark.time(result.collect)
Time taken: 146 ms
res5: Array[org.apache.spark.sql.Row] = Array([2888.906172169392])

...
```

# GPU:
```
scala> val result = df.selectExpr("std(value)")
result: org.apache.spark.sql.DataFrame = [std(CAST(value AS DOUBLE)): double]

scala> spark.time(result.collect)
23/04/28 21:05:13 WARN GpuOverrides: 
*Exec <HashAggregateExec> will run on GPU
  *Expression <AggregateExpression> std(cast(value#0 as double)) will run on GPU
    *Expression <StddevSamp> std(cast(value#0 as double)) will run on GPU
      *Expression <Cast> cast(value#0 as double) will run on GPU
  *Expression <Alias> std(cast(value#0 as double))#10 AS std(CAST(value AS DOUBLE))#27 will run on GPU
  *Exec <ShuffleExchangeExec> will run on GPU
    *Partitioning <SinglePartition$> will run on GPU
    *Exec <HashAggregateExec> will run on GPU
      *Expression <AggregateExpression> partial_std(cast(value#0 as double)) will run on GPU
        *Expression <StddevSamp> std(cast(value#0 as double)) will run on GPU
          *Expression <Cast> cast(value#0 as double) will run on GPU
      *Exec <FileSourceScanExec> will run on GPU

Time taken: 1915 ms                                                             
res0: Array[org.apache.spark.sql.Row] = Array([2888.9061721694384])

scala> spark.time(result.collect)
Time taken: 54 ms
res1: Array[org.apache.spark.sql.Row] = Array([2888.9061721694384])

scala> spark.time(result.collect)
Time taken: 43 ms
res2: Array[org.apache.spark.sql.Row] = Array([2888.9061721694384])

...
```

# GPU with 1 partition:
```
scala> val result = df.repartition(1).selectExpr("std(value)")
result: org.apache.spark.sql.DataFrame = [std(CAST(value AS DOUBLE)): double]

scala> spark.time(result.collect)
23/04/28 21:17:09 WARN GpuOverrides: 
*Exec <HashAggregateExec> will run on GPU
  *Expression <AggregateExpression> std(cast(value#0 as double)) will run on GPU
    *Expression <StddevSamp> std(cast(value#0 as double)) will run on GPU
      *Expression <Cast> cast(value#0 as double) will run on GPU
  *Expression <Alias> std(cast(value#0 as double))#10 AS std(CAST(value AS DOUBLE))#27 will run on GPU
  *Exec <HashAggregateExec> will run on GPU
    *Expression <AggregateExpression> partial_std(cast(value#0 as double)) will run on GPU
      *Expression <StddevSamp> std(cast(value#0 as double)) will run on GPU
        *Expression <Cast> cast(value#0 as double) will run on GPU
    *Exec <ShuffleExchangeExec> will run on GPU
      *Partitioning <RoundRobinPartitioning> will run on GPU
      *Exec <FileSourceScanExec> will run on GPU

Time taken: 1899 ms                                                             
res0: Array[org.apache.spark.sql.Row] = Array([2888.9061721694393])

scala> spark.time(result.collect)
Time taken: 71 ms
res1: Array[org.apache.spark.sql.Row] = Array([2888.9061721694393])

scala> spark.time(result.collect)
Time taken: 65 ms
res2: Array[org.apache.spark.sql.Row] = Array([2888.9061721694393])

...
```

**Comment by ttnghia** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala [hunk 1]
```diff
@@ -444,33 +444,86 @@ object CudfNthLikeAggregate {
...
   override val name: String = "CudfM2"
   override def dataType: DataType = DoubleType
 }
 
 class CudfMergeM2 extends CudfAggregate {
-  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar = _ =>
-    throw new UnsupportedOperationException("CudfMergeM2 aggregation is not supported in reduction")
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar =
+    (col: cudf.ColumnVector) => {
+      withResource(col.copyToHost()) { hcv =>
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:492]: nit: Would be nice to have a NVTX Range here so we can see in traces how long this is taking.

**Review by revans2 — APPROVED** [review]
Just a nit

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala [hunk 2]
```diff
@@ -444,33 +444,86 @@ object CudfNthLikeAggregate {
  * In the future, this aggregate class should be removed and the mean values should be
```
Comment (abellina) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:444]: @ttnghia I don't know if this comment applies anymore. Do you see averages coming from libcudf's M2? If not, mind updating the comment (or removing it)?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala [hunk 3]
```diff
@@ -444,33 +444,86 @@ object CudfNthLikeAggregate {
...
 class CudfM2 extends CudfAggregate {
-  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar = _ =>
-    throw new UnsupportedOperationException("CudfM2 aggregation is not supported in reduction")
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar =
+    (col: cudf.ColumnVector) => {
+      val count = col.getRowCount - col.getNullCount
+      if (count == 0) {
+        Scalar.fromDouble(0.0)
+      } else {
+        withResource(col.sum(DType.FLOAT64)) {sum =>
```
Comment (abellina) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:474]: nit
```suggestion
        withResource(col.sum(DType.FLOAT64)) { sum =>
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala [hunk 4]
```diff
@@ -444,33 +444,86 @@ object CudfNthLikeAggregate {
...
-    throw new UnsupportedOperationException("CudfMergeM2 aggregation is not supported in reduction")
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar =
+    (col: cudf.ColumnVector) => {
+      withResource(col.copyToHost()) { hcv =>
+        withResource(hcv.getChildColumnView(0)) { partialN =>
+          withResource(hcv.getChildColumnView(1)) { partialMean =>
+            withResource(hcv.getChildColumnView(2)) { partialM2 =>
+              var mergeN : Integer = 0
+              var mergeMean : Double = 0.0
+              var mergeM2 : Double = 0.0
```
Comment (abellina) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:498]: nit,
```suggestion
              var mergeN: Integer = 0
              var mergeMean: Double = 0.0
              var mergeM2: Double = 0.0
```

**Review by abellina — APPROVED** [review]
minor comments only

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala (hunk 2 above)
Comment (ttnghia) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:444]: This is still relevant. I couldn't get libcudf to output the desired output to remove it.

**Comment by ttnghia** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala [hunk 5]
```diff
@@ -489,13 +489,16 @@ class CudfM2 extends CudfAggregate {
 class CudfMergeM2 extends CudfAggregate {
   override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar =
     (col: cudf.ColumnVector) => {
-      withResource(col.copyToHost()) { hcv =>
+      val hcv = withResource(new NvtxRange("device-to-host", NvtxColor.ORANGE)) { _ =>
```
Comment (abellina) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:492]: I think the idea was to put the whole body of this reduction in a range, that way we can account for all of the time it takes (including the cpu loop)

Comment (ttnghia) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:492]: Whole body is already wrapped in another `NvtxRange`: https://github.com/NVIDIA/spark-rapids/blob/54f77b084a6b23b68af5e980d0161cf8f4ef1f1a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala#L320

Comment (ttnghia) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:492]: Anyway I'll put whole this body in `NvtxRange` too. That top level `NvtxRange` is for all reduction ops.

**Comment by ttnghia** [comment]
build

**Review by abellina — APPROVED** [review]

**Pull request merged**